storage/Devices/DeviceImpl.cc

```diff
@@ -283,7 +290,35 @@ namespace storage
     void
     Device::Impl::do_delete() const
     {
-	throw std::logic_error("stub function called");
+	ST_THROW(LogicException("stub function called"));
```

Comment:
Some devices do not support do_delete, e.g. disks. So making do_delete pure virtual would just require to reimplement it in the derived classes and likely throw an exception there.